### ISSUE TITLE
[Bug]: Incorrect type narrowing for union

### ISSUE THREAD

**Issue by rdulmina** [ballerina-platform/ballerina-lang#38503]
### Description

Consider the bellow sample

```bal
public type Utc [int, decimal];
public type ValueType Utc|record {}|byte[];

function f32(ValueType x) {
    if x is byte[]|record {} {
       
    } else {
        byte[] _ = x; // incompatible types: expected 'byte[]', found '([int,decimal]|byte[])'
    }
}
```

Since if x is `byte[]` it will go to the if block, the type of x in the else block should not contain `byte[]` it should be only `Utc`. Also, note that in the error message the type name information`(Utc)` is missing.

Also, the type name `Utc` is correctly given for the bellow sample in `2201.3.0`. But after #24078 it will also lose the type name information. Need to fix this as well.
```bal
public type Utc record & [int, decimal];
public type ValueType Utc|record {}|byte[];

function f32(ValueType x) {
    if x is byte[]|record {} {
       
    } else {
        byte[] _ = x; // incompatible types: expected 'byte[]', found '[int,decimal] & readonly'
    }
}
```

### Related area

-> Compilation

**Comment by MaryamZi** [comment]
> Consider the bellow sample
> 
> ```ballerina
> public type Utc [int, decimal];
> public type ValueType Utc|record {}|byte[];
> 
> function f32(ValueType x) {
>     if x is byte[]|record {} {
>        
>     } else {
>         byte[] _ = x; // incompatible types: expected 'byte[]', found '([int,decimal]|byte[])'
>     }
> }
> ```
> 
> Since if x is `byte[]` it will go to the if block, the type of x in the else block should not contain `byte[]` it should be only `Utc`. Also, note that in the error message the type name information`(Utc)` is missing.

This is not a bug. It is the expected behaviour according to the current rules, since `x` allows mutable values.

According to the current rules, although jBallerina doesn't allow it yet, the following is a valid argument for `x`, which means within the else block we can't guarantee the list to be a `byte[]`.

```ballerina
public function main() {
    [int, byte|decimal] x = [1, 2];
    f32(x);
}
```

See https://github.com/ballerina-platform/ballerina-lang/issues/35704#issuecomment-1091067067 also.

This will change with proposed changes to narrowing/typing though. But the current behaviour is correct according to the current spec.

**Comment by MaryamZi** [comment]
> Also, the type name `Utc` is correctly given for the bellow sample in `2201.3.0`. But after #24078 it will also lose the type name information. Need to fix this as well.
> 
> ```ballerina
> public type Utc record & [int, decimal];
> public type ValueType Utc|record {}|byte[];
> 
> function f32(ValueType x) {
>     if x is byte[]|record {} {
>        
>     } else {
>         byte[] _ = x; // incompatible types: expected 'byte[]', found '[int,decimal] & readonly'
>     }
> }
> ```

Let's create a separate issue for this.
